lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -25,21 +25,35 @@
 using namespace glow;
 
 ExecutionEngine::ExecutionEngine(BackendKind backendKind)
-    : backend_(createBackend(backendKind)) {}
+    : backend_(createBackend(backendKind)) {
+  firstRun_ = true;
+}
 
 /// Set the code generator kind to \p backendKind.
 void ExecutionEngine::setBackend(BackendKind backendKind) {
+  if (function_) {
```

Comment:
You repeat the following piece of code 3 times in your PR:

```c++
  if (function_) {
    function_->tearDownRuns();
  }
```
May be move it into a helper function?

Reply:
Another thought: May be the destructor of `function_` should always call `tearDownRuns` to ensure that all resources are  always freed when the function is deleted?  It should be possible to call `tearDownRuns` multiple times on the same function, so that both the destructor and the user of the `CompiledFunction` can do it. The subsequent calls of `tearDownRuns` should see that there is nothing to release and simply return.

Reply:
Maybe you can put it as customized deleter for the `unique_ptr`?

Reply:
On thinking about this more, I think it makes more sense to add a setup flag to compiledFunction and check that flag when setupRuns is called. Similarly call teardownRuns from the function's dtor. This way all the logic is contained within the function. Also this would make setupRuns and tearDownRuns safe to call multiple times.

Reply:
The small problem is: you can easily control the `ExecutionEngine`, but `CompiledFunction` is just an interface currently and thus you have to rely on the backends to correctly implement the logic you described above. And each backend would need to implement it correctly. May be some common logic for this could be somehow provided by the `CompiledFunction`, so that backends have less work to do?

Also, this setup flag is about managing runtime resources. And `CompiledFunction` is currently handling both the compilation and runtime aspects, but this logic will probably be separated in the future, or?

Reply:
This is probably a stop gap measure. Once it's implemented I expect the DeviceManager/DagExecutor to handle this sort of logic. Certain parts will flow forward I think. I like the idea of a requirement of the interface is that setupRuns and tearDownRuns must be capable of being called multiple times safely.